Make sslConfig agnostic of ssl-enabled DCs and enable router ssl tests

pnarayanan opened the pull request:
- Remove ssl-enabled DC parameter from sslConfig and rely on the one in
the ClusterMapConfig (which is the right place for it).
- Reenable Router-Server SSL tests.

Fixes #447 and #448

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java:
```diff
@@ -56,6 +51,32 @@ private void populatePorts(ArrayList<Port> ports) {
 
+  /**
+   * Set the datacenters to which ssl is enabled. If the datacenter on which this datanode resides is
+   * part of the datacenters to which ssl is enabled, then this datanode will always return an SSL port in
```
is there a way (or necessity) to ensure the other way round? i.e. if a node is in a DC that is ssl enabled for everyone else, then it should get an SSL port for any node that it tries to connect to.

Putting it another way, is there a way to customize this clustermap per DC? The old `getPortToConnectTo(List<String>)` provided that flexibility.

pnarayanan replied:
customizing is hard when the same clustermap is used by all nodes. The clustermap gets initialized with a config and that config has to be different for the clustermap of each node. For the main code, this makes things cleaner. In tests it would be cumbersome to create too many clustermaps unfortunately.

pnarayanan replied:
the old way provided the flexibility only in tests.

nsivabalan replied:
@vgkholla : Can you elaborate a bit on your comment ? Do you mean to suggest having different clustermaps for different DCs. For instance, for DC1, the ssl list is (DC2, DC3), for DC2, the ssl list is (DC1,DC3) and for DC3 ssl list is (DC1, DC2). Are you asking why dont we do this in our tests and we just have one clustermap for all which has (DC1, D2, DC3) ?

vgkholla replied:
@nsivabalan yup - or some way for the clustermap to figure out which DC the user of the clustermap is in. This is pretty hard to do from what I can see of the code (and from @pnarayanan's comment above)  - but would have been a nice to have.

vgkholla commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerSSLTest.java:
```diff
@@ -36,20 +41,41 @@
+  @After
+  public void after() {
+    for (MockDataNodeId m : clusterMap.getDataNodes()) {
+      if (m.sslPortWasRequested()) {
```
This seems like a weak check that might not hold if we fix some things I think.

To elaborate, it seems like the `Router` is the only one using `getPortToConnectTo()` currently. But this need not always stay true. For e.g. [this](https://github.com/linkedin/ambry/blob/master/ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java#L374) looks like duplicate code to me and this function can easily be discarded and  the single call to this function can be replaced by `getPortToConnectTo()`. (@nsivabalan - could you comment on the feasibility of removing the linked function? It looks superfluous to me).

I'm not saying that the change to `ReplicationManager` has to go into this patch (it would be great if it could!), but in the case of this test, since it also brings up servers, making that replacement would make this check invalid. What do you think?

pnarayanan replied:
Made the change in the `ReplicationManager` as well.

This check is weak but it is really to ensure that SSL tests are indeed doing what they intend to do, as that wasn't happening earlier and it went undetected.

The check for whether SSL is being exercised when asked to do SSL is not really being done in the integration tests. I think it would be hard to do in the integration tests, but unit tests should over that.

nsivabalan replied:
I guess we need something of this sort to confirm that ssl was the way the communication happened. But I am concerned about having it in after clause. I mean, how can we ensure that all of the tests will involve every data node. If in future we add tests which doesn't poll every data node this check would fail right?

pnarayanan replied:
@nsivabalan @vgkholla's comment is that this check does not necessarily ensure that the router communicated over SSL to the datanodes - because the ssl ports could be requested by the servers as part of replication - and in these tests they all share the clustermap.

pnarayanan commented on ambry-network/src/test/java/com.github.ambry.network/BlockingChannelConnectionPoolTest.java:
```diff
@@ -377,7 +387,7 @@ public void testBlockingChannelConnectionPool()
     ConnectionPool connectionPool =
         new BlockingChannelConnectionPool(new ConnectionPoolConfig(new VerifiableProperties(props)), sslConfig,
-            new MetricRegistry());
+            plainTextClusterMapConfig, new MetricRegistry());
```
the other test is in disabled state due to errors - just made the change anyway to ensure that we do the right thing when we get around to fixing that.

pnarayanan commented on ambry-server/src/integration-test/java/com.github.ambry.server/MockCluster.java:
```diff
@@ -51,27 +51,31 @@
       }
-    } catch (InstantiationException e) {
+    } catch (InstantiationException e)
+
```
fixed.

pnarayanan commented on config/frontend.properties:
```diff
@@ -15,6 +15,6 @@ rest.server.blob.storage.service.factory=com.github.ambry.frontend.AmbryBlobStor
 
 # router
 router.hostname=localhost
-router.datacenter.name=Datacenter
```
fixed.

pnarayanan commented:
About your other comments - `ToolsUtil.createSSLProperties()` is already taken care of.
Got rid of the deprecated method - that was always the plan, but required changes in tools and such.

pnarayanan commented:
@vgkholla Updated. Built and tested on OS X and Linux.

cgtz commented on config/frontend.ssl.properties:
```diff
@@ -17,7 +17,6 @@ rest.server.blob.storage.service.factory=com.github.ambry.frontend.AmbryBlobStor
 # router
 router.hostname=localhost
 router.datacenter.name=Datacenter
```
This should stay as "Datacenter", right? Since the other one was changed from "Data-Center" to "Datacenter."

nsivabalan commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java:
```diff
@@ -56,6 +51,32 @@ private void populatePorts(ArrayList<Port> ports) {
+   * Get the status for whether an SSL port was requested from this datanode.
+   * @return true if SSL port was requested; false otherwise.
+   */
+  public boolean sslPortWasRequested() {
```
why is this required ?

nsivabalan commented on ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java:
```diff
@@ -288,7 +287,7 @@ public ReplicationManager(ReplicationConfig replicationConfig, SSLConfig sslConf
       this.metricRegistry = metricRegistry;
       this.dataNodeRemoteReplicaInfosPerDC = new HashMap<String, DataNodeRemoteReplicaInfos>();
-      this.sslEnabledDatacenters = Utils.splitString(sslConfig.sslEnabledDatacenters, ",");
+      this.sslEnabledDatacenters = Utils.splitString(clusterMapConfig.clusterMapSslEnabledDatacenters, ",");
```
those will be taken as DCname (which is considered to have a whitespace). So, may not work as expected

pnarayanan commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java:
```diff
@@ -27,8 +27,8 @@
   String datacenter;
-  long rackId = -1;
   ArrayList<String> sslEnabledDataCenters = new ArrayList<String>();
+  boolean sslPortRequested = false;
```
fixed

pnarayanan commented on ambry-network/src/test/java/com.github.ambry.network/TestSSLUtils.java:
```diff
@@ -202,7 +201,7 @@ public static void addSSLProperties(Properties props, String sslEnabledDatacente
+    props.put("clustermap.ssl.enabled.datacenters", sslEnabledDatacenters);
   }
 
   /**
```
fixed

pnarayanan commented:
@vgkholla @cgtz @nsivabalan addressed comments, and made changes to authoritatively determine if router communicates over SSL or not. Please take a look at the latest commit.

vgkholla commented:
LGTM. Please verify build and formatting. Thanks for making this change ! Avoids a lot of confusion.

pnarayanan commented:
Had made a tiny change after testing out everything, which had a bug that fails the test. Just pushed the fix.
Formatted; built and tested on Linux and OS X.

cgtz commented:
LGTM

I can merge after @nsivabalan takes a look
